File: common/src/main/java/dev/mayaqq/estrogen/registry/EstrogenParticles.java
package dev.mayaqq.estrogen.registry;

import com.teamresourceful.resourcefullib.common.registry.ResourcefulRegistries;
import com.teamresourceful.resourcefullib.common.registry.ResourcefulRegistry;
import dev.mayaqq.estrogen.Estrogen;
import net.minecraft.core.particles.ParticleType;
import net.minecraft.core.particles.SimpleParticleType;
import net.minecraft.core.registries.BuiltInRegistries;
import net.minecraft.core.registries.Registries;
import uwu.serenity.critter.api.entry.RegistryEntry;
import uwu.serenity.critter.stdlib.Registrar;

public class EstrogenParticles {
    public static final Registrar<ParticleType<?>> PARTICLES = Registrar.create(Estrogen.MOD_ID, Registries.PARTICLE_TYPE);

    public static final RegistryEntry<SimpleParticleType> DASH = PARTICLES.<SimpleParticleType>entry("dash", () -> new SimpleParticleType(true) {}).register();
    public static final RegistryEntry<SimpleParticleType> MOTH_FUZZ = PARTICLES.<SimpleParticleType>entry("moth_fuzz", () -> new SimpleParticleType(true) {}).register();
}

File: src/common/main/kotlin/dev/mayaqq/estrogen/content/EstrogenParticles.kt
package dev.mayaqq.estrogen.content

import dev.mayaqq.estrogen.Estrogen
import dev.mayaqq.estrogen.MOD_ID
import dev.mayaqq.estrogen.client.content.particles.DashTrailParticle
import dev.mayaqq.estrogen.client.content.particles.MothFuzzParticle
import dev.mayaqq.estrogen.content.particles.DashTrailParticleOptions
import invoke.kitty.kritter.registry.api.Registrar
import invoke.kitty.kritter.registry.misc.particleType
import net.minecraft.client.particle.ParticleProvider
import net.minecraft.core.particles.ParticleType
import net.minecraft.core.particles.SimpleParticleType
import net.minecraft.core.registries.Registries

//@EventSubscriber(env = [Environment.CLIENT])
object EstrogenParticles : Registrar<ParticleType<*>> by Registrar(MOD_ID, Registries.PARTICLE_TYPE) {

    val MothFuzz: SimpleParticleType by particleType("moth_fuzz") {
        provider { sprites -> ParticleProvider { _, cilientLevel, x, y, z, _, _, _ -> MothFuzzParticle(clientLevel, x, y, z, sprites) } }
    }i

    val DashTrail: CynosureParticleType<DashTrailParticleOptions> by particleType(
        "dash_trail",
        DashTrailParticleOptions.CODEC,
        DashTrailParticleOptions.NETWORK_CODEC
    ) {
        provider(::DashTrailParticle)
        overrideLimiter = true
    }

    // TODO: Implement dream block ripples from walking on punching
    val DreamBlockRipple: SimpleParticleType by particleType("dream_ripple") {
        overrideLimiter = true
    }

}